File: sdk/lua/darkauth.lua
--[[
    DARK-AUTH Official Lua SDK v2.0.0
    Compatible with FiveM, Roblox, and Standalone Lua (luasocket/curl).
]]

local DarkAuth = {}
DarkAuth.__index = DarkAuth

local SDK_VERSION = "2.0.0"

function DarkAuth.new(appId, secret, apiUrl, version)
    local self = setmetatable({}, DarkAuth)
    self.appId = appId
    self.secret = secret
    self.version = version or SDK_VERSION
    self.apiUrl = (apiUrl or ""):gsub("/+$", "")
    self.sessionToken = nil
    self.hwid = self:_generateHWID()
    self.user = nil
    self.licenseInfo = nil
    return self
end

function DarkAuth:_generateHWID()
    if _G.Citizen then
        -- FiveM environment
        local identifiers = {}
        local source = GetPlayerName(PlayerId()) or "fivem"
        table.insert(identifiers, source)
        table.insert(identifiers, tostring(GetPlayerServerId(PlayerId())))
        local raw = table.concat(identifiers, ":")
        return self:_sha256(raw)
    end
    if game and game:GetService then
        -- Roblox environment
        local HttpService = game:GetService("HttpService")
        local Players = game:GetService("Players")
        local player = Players.LocalPlayer
        local raw = tostring(player and player.UserId or 0) .. ":" .. tostring(player and player.Name or "unknown")
        return self:_sha256(raw)
    end
    -- Standalone
    local hostname = ""
    if os and os.hostname then
        hostname = os.hostname()
    end
    local user = ""
    if os and os.getenv then
        user = os.getenv("USER") or os.getenv("USERNAME") or "lua"
    end
    return self:_sha256(hostname .. ":" .. user .. ":" .. tostring(os.time()))
end

function DarkAuth:_sha256(input)
    -- FiveM has Citizen.ValidateText, Roblox has crypto, standalone uses openssl
    if _G.Citizen then
        return exports["darkauth"]:sha256(input) or self:_fallbackHash(input)
    end
    if crypto and crypto.sha256 then
        return crypto.sha256(input):lower()
    end
    return self:_fallbackHash(input)
end

function DarkAuth:_fallbackHash(input)
    -- Simple fallback using MD5-like approach (not cryptographically secure)
    local hash = 0
    for i = 1, #input do
        hash = ((hash * 31) + string.byte(input, i)) % 2147483647
    end
    return string.format("%064x", hash)
end

function DarkAuth:_url(endpoint)
    return self.apiUrl .. "/api/v2/" .. endpoint:gsub("^/+", "")
end

function DarkAuth:_httpPost(endpoint, data)
    local url = self:_url(endpoint)

    -- FiveM environment
    if PerformHttpRequest then
        local p = promise.new()
        PerformHttpRequest(url, function(statusCode, responseText, headers)
            local ok, decoded = pcall(json.decode, responseText)
            if ok then
                p:resolve(decoded)
            else
                p:resolve({ success = false, message = "JSON decode failed" })
            end
        end, "POST", json.encode(data), { ["Content-Type"] = "application/json" })
        local res = Citizen.Await(p)
        return res
    end

    -- Roblox environment
    if game and game:GetService then
        local HttpService = game:GetService("HttpService")
        local success, response = pcall(function()
            return HttpService:PostAsync(url, HttpService:JSONEncode(data), Enum.HttpContentType.ApplicationJson)
        end)
        if success then
            return HttpService:JSONDecode(response)
        else
            return { success = false, message = tostring(response) }
        end
    end

    -- Standalone fallback via curl
    local jsonParts = {}
    for k, v in pairs(data) do
        table.insert(jsonParts, string.format('"%s":"%s"', tostring(k), tostring(v)))
    end
    local jsonStr = "{" .. table.concat(jsonParts, ",") .. "}"

    local cmd = string.format('curl -s -X POST "%s" -H "Content-Type: application/json" -d \'%s\'', url, jsonStr)
    local handle = io.popen(cmd)
    if not handle then
        return { success = false, message = "curl failed" }
    end
    local result = handle:read("*a")
    handle:close()

    -- Parse result manually
    local success = result:find('"success":true') ~= nil
    local sessionToken = result:match('"session_token":"([^"]+)"')
    local value = result:match('"value":"([^"]+)"')
    local message = result:match('"message":"([^"]+)"')
    local code = result:match('"code":"([^"]+)"')

    return {
        success = success,
        session_token = sessionToken,
        value = value,
        message = message,
        code = code,
        raw = result
    }
end

function DarkAuth:_httpGet(endpoint, params)
    local url = self:_url(endpoint)
    local queryString = {}
    for k, v in pairs(params) do
        table.insert(queryString, k .. "=" .. tostring(v))
    end
    if #queryString > 0 then
        url = url .. "?" .. table.concat(queryString, "&")
    end

    -- FiveM
    if PerformHttpRequest then
        local p = promise.new()
        PerformHttpRequest(url, function(statusCode, responseText)
            local ok, decoded = pcall(json.decode, responseText)
            p:resolve(ok and decoded or { success = false, message = "JSON decode failed" })
        end, "GET")
        return Citizen.Await(p)
    end

    -- Roblox
    if game and game:GetService then
        local HttpService = game:GetService("HttpService")
        local success, response = pcall(function()
            return HttpService:GetAsync(url)
        end)
        if success then
            return HttpService:JSONDecode(response)
        end
        return { success = false, message = tostring(response) }
    end

    -- Standalone
    local handle = io.popen('curl -s "' .. url .. '"')
    if not handle then return { success = false, message = "curl failed" } end
    local result = handle:read("*a")
    handle:close()

    local success = result:find('"success":true') ~= nil
    local messages = {}
    for msg in result:gmatch('"message":"([^"]*)"') do
        table.insert(messages, { message = msg })
    end
    return { success = success, messages = messages, raw = result }
end

function DarkAuth:init(hash)
    local data = {
        app_id = self.appId,
        secret = self.secret,
        version = self.version
    }
    if hash then data.hash = hash end

    local res = self:_httpPost("/init", data)
    if res.success then
        self.sessionToken = res.session_token
    end
    return res
end

function DarkAuth:license(key)
    if not self.sessionToken then self:init() end
    local res = self:_httpPost("/license", {
        session_token = self.sessionToken,
        key = key:match("^%s*(.-)%s*$"),
        hwid = self.hwid
    })
    self.licenseInfo = res
    self.user = res.user
    return res
end

function DarkAuth:login(username, password)
    if not self.sessionToken then self:init() end
    local res = self:_httpPost("/login", {
        session_token = self.sessionToken,
        username = username,
        password = password,
        hwid = self.hwid
    })
    self.user = res.user
    return res
end

function DarkAuth:register(username, password, key)
    if not self.sessionToken then self:init() end
    local res = self:_httpPost("/register", {
        session_token = self.sessionToken,
        username = username,
        password = password,
        key = key:match("^%s*(.-)%s*$"),
        hwid = self.hwid
    })
    self.user = res.user
    return res
end

function DarkAuth:check()
    if not self.sessionToken then return false end
    local res = self:_httpPost("/check", { session_token = self.sessionToken })
    return res.success == true
end

function DarkAuth:getVar(name)
    if not self.sessionToken then self:init() end
    local res = self:_httpPost("/var/get", {
        session_token = self.sessionToken,
        name = name
    })
    return res.value
end

function DarkAuth:setVar(name, value)
    if not self.sessionToken then self:init() end
    return self:_httpPost("/var/set", {
        session_token = self.sessionToken,
        name = name,
        value = tostring(value)
    })
end

function DarkAuth:log(message, level)
    if not self.sessionToken then self:init() end
    return self:_httpPost("/log", {
        session_token = self.sessionToken,
        message = message,
        level = level or "INFO"
    })
end

function DarkAuth:resetHWID(key)
    if not self.sessionToken then self:init() end
    return self:_httpPost("/hwid/reset", {
        session_token = self.sessionToken,
        key = key
    })
end

function DarkAuth:getChat(channel)
    if not self.sessionToken then self:init() end
    local res = self:_httpGet("/chat", {
        session_token = self.sessionToken,
        channel = channel or "general"
    })
    return res.messages or {}
end

function DarkAuth:sendChat(sender, message, channel)
    if not self.sessionToken then self:init() end
    return self:_httpPost("/chat", {
        session_token = self.sessionToken,
        channel = channel or "general",
        sender = sender,
        message = message
    })
end

return DarkAuth

File: sdk/csharp/DarkAuth.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Management;
using System.Net.Http;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Threading.Tasks;

namespace DarkAuthSdk
{
    /// <summary>DARK-AUTH Official C# SDK v2.0.0</summary>
    public class DarkAuth : IDisposable
    {
        public string AppId { get; }
        public string Secret { get; }
        public string Version { get; }
        public string ApiUrl { get; }
        public string SessionToken { get; private set; }
        public string Hwid { get; }
        public Dictionary<string, object> User { get; private set; }
        public Dictionary<string, object> LicenseInfo { get; private set; }

        private static readonly HttpClient _httpClient = new HttpClient { Timeout = TimeSpan.FromSeconds(15) };

        private static readonly JsonSerializerOptions _jsonOptions = new JsonSerializerOptions
        {
            PropertyNamingPolicy = JsonNamingPolicy.SnakeCaseLower,
            DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
        };

        public DarkAuth(string appId, string secret, string apiUrl, string version = "2.0.0")
        {
            AppId = appId;
            Secret = secret;
            Version = version;
            ApiUrl = apiUrl.TrimEnd('/');
            Hwid = GenerateHwid();
        }

        /// <summary>Generate a hardware fingerprint from system information.</summary>
        public static string GenerateHwid()
        {
            try
            {
                var sb = new StringBuilder();
                sb.Append(Environment.MachineName);
                sb.Append(':');
                sb.Append(Environment.UserName);

                try
                {
                    using var searcher = new ManagementObjectSearcher("SELECT UUID FROM Win32_ComputerSystemProduct");
                    foreach (var obj in searcher.Get())
                    {
                        sb.Append(':');
                        sb.Append(obj["UUID"]?.ToString() ?? "");
                    }
                }
                catch { }

                using var sha = SHA256.Create();
                var bytes = sha.ComputeHash(Encoding.UTF8.GetBytes(sb.ToString()));
                return BitConverter.ToString(bytes).Replace("-", "").ToLowerInvariant();
            }
            catch
            {
                return "csharp_default_hwid";
            }
        }

        /// <summary>Compute SHA-256 hash of a file for integrity checking.</summary>
        public static string ComputeFileHash(string filePath)
        {
            using var sha = SHA256.Create();
            using var stream = File.OpenRead(filePath);
            var bytes = sha.ComputeHash(stream);
            return BitConverter.ToString(bytes).Replace("-", "").ToLowerInvariant();
        }

        private string Url(string endpoint)
        {
            return $"{ApiUrl}/api/v2/{endpoint.TrimStart('/')}";
        }

        private async Task<JsonElement> PostAsync(string endpoint, object data)
        {
            var json = JsonSerializer.Serialize(data, _jsonOptions);
            var content = new StringContent(json, Encoding.UTF8, "application/json");
            var response = await _httpClient.PostAsync(Url(endpoint), content);
            var resString = await response.Content.ReadAsStringAsync();
            var doc = JsonDocument.Parse(resString);
            var root = doc.RootElement;

            if (root.TryGetProperty("success", out var success) && !success.GetBoolean())
            {
                var msg = root.TryGetProperty("message", out var m) ? m.GetString() : "Request failed";
                var code = root.TryGetProperty("code", out var c) ? c.GetString() : "ERROR";
                throw new Exception($"[{code}] {msg}");
            }
            return root;
        }

        private async Task<JsonElement> GetAsync(string endpoint, Dictionary<string, string> queryParams)
        {
            var qs = string.Join("&", queryParams.Select(kv => $"{kv.Key}={Uri.EscapeDataString(kv.Value)}"));
            var url = string.IsNullOrEmpty(qs) ? Url(endpoint) : $"{Url(endpoint)}?{qs}";
            var response = await _httpClient.GetAsync(url);
            var resString = await response.Content.ReadAsStringAsync();
            return JsonDocument.Parse(resString).RootElement;
        }

        /// <summary>Initialize session with the DARK-AUTH API.</summary>
        public async Task<JsonElement> InitAsync(string binaryPath = null)
        {
            var payload = new Dictionary<string, object>
            {
                ["app_id"] = AppId,
                ["secret"] = Secret,
                ["version"] = Version,
            };
            if (!string.IsNullOrEmpty(binaryPath))
                payload["hash"] = ComputeFileHash(binaryPath);

            var res = await PostAsync("/init", payload);
            if (res.TryGetProperty("session_token", out var token))
                SessionToken = token.GetString();
            return res;
        }

        /// <summary>Authenticate with a license key.</summary>
        public async Task<JsonElement> LicenseAsync(string key)
        {
            if (string.IsNullOrEmpty(SessionToken)) await InitAsync();
            var res = await PostAsync("/license", new Dictionary<string, string>
            {
                ["session_token"] = SessionToken,
                ["key"] = key.Trim(),
                ["hwid"] = Hwid,
            });
            LicenseInfo = res.EnumerateObject().ToDictionary(p => p.Name, p => (object)(p.Value.ToString()));
            if (res.TryGetProperty("user", out var userProp))
                User = userProp.EnumerateObject().ToDictionary(p => p.Name, p => (object)(p.Value.ToString()));
            return res;
        }

        /// <summary>Log in with username and password.</summary>
        public async Task<JsonElement> LoginAsync(string username, string password)
        {
            if (string.IsNullOrEmpty(SessionToken)) await InitAsync();
            var res = await PostAsync("/login", new Dictionary<string, string>
            {
                ["session_token"] = SessionToken,
                ["username"] = username,
                ["password"] = password,
                ["hwid"] = Hwid,
            });
            if (res.TryGetProperty("user", out var userProp))
                User = userProp.EnumerateObject().ToDictionary(p => p.Name, p => (object)(p.Value.ToString()));
            return res;
        }

        /// <summary>Register a new account with a license key.</summary>
        public async Task<JsonElement> RegisterAsync(string username, string password, string key)
        {
            if (string.IsNullOrEmpty(SessionToken)) await InitAsync();
            var res = await PostAsync("/register", new Dictionary<string, string>
            {
                ["session_token"] = SessionToken,
                ["username"] = username,
                ["password"] = password,
                ["key"] = key.Trim(),
                ["hwid"] = Hwid,
            });
            if (res.TryGetProperty("user", out var userProp))
                User = userProp.EnumerateObject().ToDictionary(p => p.Name, p => (object)(p.Value.ToString()));
            return res;
        }

        /// <summary>Check if the current session is still valid.</summary>
        public async Task<bool> CheckAsync()
        {
            if (string.IsNullOrEmpty(SessionToken)) return false;
            try
            {
                var res = await PostAsync("/check", new Dictionary<string, string> { ["session_token"] = SessionToken });
                return res.TryGetProperty("success", out var s) && s.GetBoolean();
            }
            catch { return false; }
        }

        /// <summary>Get a cloud variable value.</summary>
        public async Task<string> GetVarAsync(string name)
        {
            var res = await PostAsync("/var/get", new Dictionary<string, string>
            {
                ["session_token"] = SessionToken,
                ["name"] = name,
            });
            return res.TryGetProperty("value", out var val) ? val.GetString() : null;
        }

        /// <summary>Set a cloud variable value.</summary>
        public async Task<JsonElement> SetVarAsync(string name, string value)
        {
            return await PostAsync("/var/set", new Dictionary<string, string>
            {
                ["session_token"] = SessionToken,
                ["name"] = name,
                ["value"] = value,
            });
        }

        /// <summary>Send a log entry to the API.</summary>
        public async Task LogAsync(string message, string level = "INFO")
        {
            await PostAsync("/log", new Dictionary<string, string>
            {
                ["session_token"] = SessionToken,
                ["message"] = message,
                ["level"] = level,
            });
        }

        /// <summary>Request a HWID reset for a key.</summary>
        public async Task<JsonElement> ResetHwidAsync(string key)
        {
            return await PostAsync("/hwid/reset", new Dictionary<string, string>
            {
                ["session_token"] = SessionToken,
                ["key"] = key,
            });
        }

        /// <summary>Get chat messages from a channel.</summary>
        public async Task<List<Dictionary<string, object>>> GetChatAsync(string channel = "general")
        {
            var res = await GetAsync("/chat", new Dictionary<string, string>
            {
                ["session_token"] = SessionToken,
                ["channel"] = channel,
            });
            if (res.TryGetProperty("messages", out var messages))
            {
                return messages.EnumerateArray()
                    .Select(m => m.EnumerateObject().ToDictionary(p => p.Name, p => (object)(p.Value.ToString())))
                    .ToList();
            }
            return new List<Dictionary<string, object>>();
        }

        /// <summary>Send a chat message.</summary>
        public async Task<JsonElement> SendChatAsync(string sender, string message, string channel = "general")
        {
            return await PostAsync("/chat", new Dictionary<string, string>
            {
                ["session_token"] = SessionToken,
                ["channel"] = channel,
                ["sender"] = sender,
                ["message"] = message,
            });
        }

        public void Dispose()
        {
            _httpClient?.Dispose();
        }
    }
}
